터미널환경설정 FE 구현

RBJH opened the pull request:
- Flux 패턴 적용
- 각 카테고리 스테퍼 구현
- 스테퍼 및 버튼 이벤트에 따라 카테고리 전환

dimsssss committed:
Merge: pull request #192 from dimsssss/feature/setting

사용자 터미널 환경 구성

RBJH committed:
Feat: 터미널 환경 API 구현

터미널 환경설정창에서 선택한 환경들을 기반으로 Dockerfile을 생성

RBJH committed:
Merge: branch 'develop' of https://github.com/connect-foundation/2019-14 into develop

RBJH committed:
Chore: CSS flex 속성 조정

CSS 변경 간 예상하지 못한 flex-direction 문제로
flex-direction속성 column 추가

RBJH committed:
Refactor: 툴 바 및 터미널 환경설정 코드 리팩토링

중복 코드 부분 제거 및 기존 조건부 렌더링 부분 비용을 고려해
css 속성으로 hidden여부를 props로 전달하는 방법으로 변경

RBJH committed:
Chore: CSS 속성 값  수정

width 적용 범위가 부적절해 각 셀들의 width 값 수정

RBJH committed:
Merge: branch 'develop' of https://github.com/connect-foundation/2019-14 into develop

RBJH commented on client/src/actions/TerminalSetting.js:
```diff
@@ -0,0 +1,50 @@
+    PL: "terminal-setting/select/pl",
+    DB: "terminal-setting/select/db",
+  },
+  UNSELECT: {
```
삭제 필요

RBJH commented on client/src/components/editor/header/TerminalImageContainer.jsx:
```diff
@@ -1,23 +1,44 @@
-const TerminalImageContainer = (props) => {
+const TerminalImageContainer = () => {
+  const { state } = useContext(TerminalSettingContext);
   const makeImageList = () => {
```
이 함수들 다 제거 여부(코드 리뷰)

ImuruKevol commented on client/src/components/editor/header/TerminalImageContainer.jsx:
```diff
@@ -1,23 +1,44 @@
+    });
+
+    return images.map((image, index) => {
+      return <FontAwesomeIcon key={index} icon={ICON[image]} />;
```
index를 키값으로 바로 쓰시면 안됩니다!

ImuruKevol commented on client/src/components/editor/terminal-setting-drawers/Stepper.jsx:
```diff
@@ -12,7 +12,8 @@ const StepperWrapper = styled.header`
   const getSteps = () => {
     return Object.values(STEP_TYPE).map((icon, index) => {
-      return <Step icon={icon} key={index} />;
+      const key = `key-${index}`;
```
단순히 key를 붙이는 것보단 의미를 담는게 더 좋을 것 같습니다.

ImuruKevol commented on client/src/components/editor/header/TerminalImageContainer.jsx:
```diff
@@ -1,23 +1,44 @@
+    let images = [];
+
+    Object.keys(state).map((element) => {
+      if (element === "OS" || element === "PL" || element === "DB")
```
이 부분은 Object.keys와 includes같은걸 이용해서 변수에 담아서 하는 것도 좋을 것 같습니다.

DrizzlingCattus commented on client/src/reducers/TerminalSetting.js:
```diff
@@ -0,0 +1,65 @@
+    return null;
+  }
+
+  const handler = terminalSettingHendler[action.type];
```
cellReducer쪽을 살펴보고 만들면 더 좋을것같아요!

DrizzlingCattus replied:
오타 Hendler -> handler로 부탁드립니다.

DrizzlingCattus commented on client/src/reducers/TerminalSetting.js:
```diff
@@ -0,0 +1,65 @@
+
+    newOs.push(action.os);
+
+    return { ...state, OS: newOs };
```
spread가 되서 얕은 복사가 되서 깊은 복사가 필요하다는 사실을 체크해주시면 감사합니다!

ImuruKevol commented on client/src/reducers/TerminalSetting.js:
```diff
@@ -0,0 +1,65 @@
+    const targetIndex = newPL.findIndex((element) => element === action.pl);
+
+    if (targetIndex < 0) newPL.push(action.pl);
+    else newPL.splice(targetIndex, 1);
```
splice는 뮤터블한 애라서 제가 이뮤터블하게 만들어놓은 splice 모듈이 있습니다.
newPL = state.PL
이렇게 하게 되면 어차피 레퍼런스 복사이기 때문에 state의 값을 직접 바꾸게 되는 셈입니다.
이런 방식은 플럭스 패턴에 매우 좋지 않습니다.

RBJH committed:
Merge: branch 'develop' of https://github.com/connect-foundation/2019-14 into develop

RBJH committed:
Merge: branch 'develop' of https://github.com/connect-foundation/2019-14 into develop

RBJH committed:
Fix: 코드 리뷰사항 반영

불필요한 코드 제거
React list & key 키 값 인덱스 대신 구체화
터미널 환경 언어에 대해 환경 자체로 변경
그에 따른 각 action, reducer, store 변수명, 상수 변경

RBJH committed:
Merge: remote-tracking branch 'upstream/develop' into develop
